contracts/hydro/src/contract.rs

```diff
@@ -676,135 +675,210 @@ fn vote(
         .add_attribute("action", "vote")
         .add_attribute("sender", info.sender.to_string());
 
-    // Get any existing vote for this sender and reverse it- this may be a vote for a different proposal (if they are switching their vote),
-    // or it may be a vote for the same proposal (if they have increased their power by locking more and want to update their vote).
-    // TODO: this could be made more gas-efficient by using a separate path with fewer writes if the vote is for the same proposal
-    let vote = VOTE_MAP.load(deps.storage, (round_id, tranche_id, info.sender.clone()));
-    if let Ok(vote) = vote {
-        // Load the proposal in the vote
-        let mut proposal = PROPOSAL_MAP.load(deps.storage, (round_id, tranche_id, vote.prop_id))?;
-
-        // Remove proposal's old power in PROPS_BY_SCORE
-        PROPS_BY_SCORE.remove(
-            deps.storage,
-            (
-                (round_id, proposal.tranche_id),
-                proposal.power.into(),
-                vote.prop_id,
-            ),
-        );
+    // Eliminate duplicate proposal and lock IDs
+    let mut proposal_ids = HashSet::new();
+    let mut lock_ids = HashSet::new();
 
-        remove_many_validator_shares_from_proposal(
-            deps.storage,
-            round_id,
-            vote.prop_id,
-            vote.time_weighted_shares
-                .iter()
-                .map(|(k, v)| (k.clone(), *v))
-                .collect::<Vec<_>>(),
-        )?;
+    let proposals_votes: Vec<ProposalToLockups> = proposals_votes
+        .into_iter()
+        .filter_map(|prop_to_lockups| {
+            // filter out non unique proposal IDs
+            if proposal_ids.insert(prop_to_lockups.proposal_id) {
+                // filter out non unique lock IDs
+                let unique_lock_ids: Vec<u64> = prop_to_lockups
+                    .lock_ids
+                    .into_iter()
+                    .filter(|lock_id| lock_ids.insert(*lock_id))
+                    .collect();
+
+                if !unique_lock_ids.is_empty() {
+                    Some(ProposalToLockups {
+                        proposal_id: prop_to_lockups.proposal_id,
+                        lock_ids: unique_lock_ids,
+                    })
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        })
+        .collect();
```

Comment:
what happens when a lock_id is mentioned twice?

Reply:
It will be taken into account only the first time it is mentioned (see line 691). Personally, I think we should completely reject such transactions, but that isn't the approach we have taken so far, and therefore I implemented the filtering.

Reply:
I think we should reject the transaction, because there is no obvious way to determine which was the intended vote if it appears twice.

Reply:
If it doesn't make the code less elegant, let's go for that approach

Reply:
Done in https://github.com/informalsystems/hydro/pull/157/commits/ece382d60f90a320440f6d84ef8270d3cc632f64